neural_structured_learning/examples/notebooks/adversarial_cnn_transfer_learning_fashionmnist.ipynb

```diff
@@ -0,0 +1,3081 @@
+    "$$adv\\_x = x + \\epsilon*\\text{sign}(\\nabla_xJ(\\theta, x, y))$$\n",
+    "\n",
+    "where \n",
+    "\n",
+    "*   adv_x : Adversarial image.\n",
+    "*   x : Original input image.\n",
+    "*   y : Original input label.\n",
+    "*   $\\epsilon$ : Multiplier to ensure the perturbations are small.\n",
+    "*   $\\theta$ : Model parameters.\n",
+    "*   $J$ : Loss.\n",
+    "\n",
+    "An intriguing property here, is the fact that the gradients are taken with respect to the input image. This is done because the objective is to create an image that maximises the loss. \n",
+    "\n",
+    "A method to accomplish this is to find how much each pixel in the image contributes to the loss value, and add a perturbation accordingly. \n",
+    "\n",
+    "This works pretty fast because it is easy to find how each input pixel contributes to the loss by using the chain rule and finding the required gradients. \n",
+    "\n",
+    "Hence, the gradients are taken with respect to the image. In addition, since the model is no longer being trained (thus the gradient is not taken with respect to the trainable variables, i.e., the model parameters), and so the model parameters remain constant. The only goal is to fool an already trained model.\n",
+    "\n",
+    "So let's try and fool our pretrained VGG19 model."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "hvIozwLu5rV5"
+   },
+   "source": [
+    "## Utility Functions for FGSM\n",
+    "\n",
+    "1. __`get_model_preds(...)`__: Helps in getting the top predicted class label and probability of an input image based on a specific trained CNN model\n",
+    "\n",
+    "2. __`generate_adverarial_pattern(...)`__: Helps in getting the gradients and the sign of the gradients w.r.t the input image and the trained CNN model\n",
+    "\n",
+    "3. __`perform_adversarial_attack_fgsm(...)`__: Create perturbations which will be used to distort the original image resulting in an adversarial image by adding epsilon to the gradient signs (can be added to gradients also) and then showcase model performance on the same"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "colab": {},
+    "colab_type": "code",
+    "id": "ecHH5CvIuas2"
+   },
+   "outputs": [],
+   "source": [
+    "def get_model_preds(input_image, class_names_map, model):\n",
+    "        preds = model.predict(input_image)\n",
+    "        top_idx = np.argsort(-preds)[0][0]\n",
+    "        top_prob = -np.sort(-preds)[0][0]\n",
+    "        top_class = np.array(class_names_map)[top_idx]\n",
+    "        return top_class, top_prob\n",
```

Comment:
Indentation is not correct (here and below in other functions). We typically use two spaces for indentation.

Reply:
got it, will correct the code indents across the notebook